PR for RVV support using splitted small chunks

zhongjuzhe opened the pull request:
This is a PR for review convenience.

zhongjuzhe committed:
PR for RVV support using splitted small chunks

zhongjuzhe committed:
add zfh string support for -march:
  user can use -march=rv64gcv_zfh to enable zfh support

zhongjuzhe committed:
add RVV vector mode including full vector mode, partial vector mode
  full tuple mode, partial tuple mode

zhongjuzhe committed:
add rvv instruction type. It is used for instruction scheduling

zhongjuzhe committed:
add zfh support in riscv.md

zhongjuzhe committed:
add strlen and strcmp and memcpy support using rvv instructions

zhongjuzhe committed:
add soft half-precision float

zhongjuzhe committed:
add predicated and constraint for RVV

zhongjuzhe committed:
add RVV vector registers

zhongjuzhe committed:
add header "riscv_vector.h" for RVV

zhongjuzhe committed:
add poly_int movement support:
  For example: (set DI reg) (const_poly_int 16 16).
  This pattern is translated into asm csrr reg, vlenb.
  For other poly_int value, we translate it into
  csrr reg, vlenb + scalar operation (shift, mul, plus).

zhongjuzhe committed:
stack manipulation && prologue handling && epilogue handling

kito-cheng commented on gcc/config/riscv/riscv.md:
```diff
@@ -154,7 +226,7 @@
 ;; shift	integer shift instructions
 ;; slt		set less than instructions
-;; imul		integer multiply 
+;; imul		integer multiply
```
Don't fix that this in this patch, maybe you could collect those clean up stuff to single patch and go upstream directly.

zhongjuzhe replied:
I fix the issue here:https://github.com/riscv-collab/riscv-gcc/pull/334/commits/15ba09286c0af8f6e23a2f9ece7a7a716002556b. Feel free to comment it

zhongjuzhe commented on gcc/config/riscv/riscv.h:
```diff
@@ -22,6 +22,7 @@ along with GCC; see the file COPYING3.  If not see
 #ifndef GCC_RISCV_H
 #define GCC_RISCV_H
 
+#include <stdbool.h>
```
I tried to remove this, but it reports error during compilation. Because I am using "bool used_vrs[MAX_ARGS_IN_VECTOR_REGISTERS]" in the CUMULATIVE_ARGS structure. This variable is used to handle function arg of RVV vector register. Do you want me to change it into "int used_vrs[MAX_ARGS_IN_VECTOR_REGISTERS]"?

kito-cheng replied:
Oh this file has included during compiling libgcc, and that's in C mode..., okay, let keep that!

zhongjuzhe replied:
> 

I saw your comments. I agree these issues should be fixed. Thanks for that. Let me finish splitting the patch and upload the rest of the codes. Then I will fix the issues one by one as you mention.

zhongjuzhe committed:
add implementation for memcpy, strlen, strcpy and strcmp

zhongjuzhe committed:
add python to generates iterators.def:
  riscv_md_parser.py this script is used to generate riscv-vector-builtins-iterators.def
  file using vector-iterators.md.
  The definition of riscv-vector-builtins-iterators.def is used by intrinsic definition
  For example:If we define DEF_RVV_FUNCTION(vsetvl, config, (1, VITER(VI, signed))....)
  the vsetvl of all machine_mode in the "VI" iterators are added into the builtins.

zhongjuzhe committed:
add vector_type and tuple_type definition using macro DEF_RVV_TYPE
  For example: DEF_RVV_TYPE (float16, fp16_type_node). This macro will
  define all data_type (vfloat16m1_t .....)
  and tuple type (vfloat16m1x2_t ......)

zhongjuzhe committed:
add RVV intrinsic framework and add RVV Configuration-Setting and Utility Functions
  which are defined as misc(The name can be changed) functions.For example:
  1.readvl. 2.vreinterpret 3.vlmul_ext 4.vlmul_trunc 5.vundefined
  6.vset 7.vget

zhongjuzhe committed:
add vlm, vsm, vle, vse intrinsics

zhongjuzhe committed:
add strided load store intrinsic (vlse and vsse)

zhongjuzhe committed:
add indexed load store (vluxei vsuxei vloxei vluxei)

zhongjuzhe committed:
add vleff load. This intrinsic is special we use gimple fold for
  the intrinsic.

zhongjuzhe committed:
add segment load store and finish all load store implementation

zhongjuzhe committed:
add load store MD patterns for last patterns(Forgot add it in the
  last commit).

zhongjuzhe committed:
add integer arithmetic intrinsics(vadd,vsub,vrsub,vwadd,vwsub,
  vext,vzext,vadc,vsbc,vmadc,vmsbc,vand,vor,vxor,vsll,vsrl,vsra,
  vmscmp,vmin,vmax,vmul,vmulh,vwmul,vdiv,vrem,vmadd,vnmsac,vmacc,
  vnmsub,vwmacc,vmerge,vmv

zhongjuzhe committed:
add chapter 12. fixed-point arithmetic intrinsics

zhongjuzhe committed:
add chapter 13. fixed-point arithmetic arithmetic

zhongjuzhe committed:
add chapter 14. vector reduction intrinsics

zhongjuzhe committed:
add chapter 15 vector mask intrinsics

zhongjuzhe committed:
add chapter 16. permutation intrinsics

zhongjuzhe committed:
add RVV intrinsic framework to riscv-builtins.cc:
  we seperate builtins into 2 types:
  1. RISCV_BUILTIN_VECTOR (RVV intrinsic)
  2. RISCV_BUILTIN_GENERAL(Normal intrinsic that original RISCV support)

zhongjuzhe committed:
add insert-vsetvli PASS. This is used for insert vsetvli instruction
  for every RVV instructions. The algorithm is following the LLVM

zhongjuzhe committed:
add #pragma riscv intrinsic "vector" support for RVV intrinsics

zhongjuzhe committed:
change all HOST_WIDE_INT into poly_int64. Because we should use
  poly_int for scalable vector

zhongjuzhe committed:
add zfh support in riscv.cc

zhongjuzhe committed:
add csrr %0,vlenb asm support for RVV

zhongjuzhe committed:
add RVV vector function arg support:
  we follows the LLVM: v0 is used to pass mask register.
  v8 - v23 are used to pass vector register

zhongjuzhe committed:
add mriscv-vector-bits compile option analysis

zhongjuzhe committed:
add mangle and floating-point target hook

zhongjuzhe committed:
add scalable vector configuration target hook

zhongjuzhe committed:
add autovectorization target hook support and
  autovectorization patterns :vec_duplicate

zhongjuzhe committed:
add vector alignment target hook

zhongjuzhe committed:
add vector initialization auto-vectorization patterns:
  vec_set vec_insert vec_init vec_extract...
  The vec_init pattern is quite complicated, the implementation is
  in the riscv.cc (riscv_vector_expand_vector_init)

zhongjuzhe committed:
add load_lanes && store lanes && gather_load && scatter_store
  (no mask) auto-vectorization patterns

zhongjuzhe committed:
add mask target hook support
  add mask load store auto-vectorization support

zhongjuzhe committed:
add auto-vectorizaiton patterns:unary operation

zhongjuzhe committed:
add auto-vectorization patterns:conversion

zhongjuzhe committed:
add auto-vectorizaiton patterns:binary operation

zhongjuzhe committed:
add auto-vectorization patterns:ternary operation

zhongjuzhe committed:
add auto-vectorization patterns:reduction operation

zhongjuzhe committed:
add sew=64 support vx support in RV32 system.

zhongjuzhe committed:
add vcond vec_cmp vcond_mask auto-vectorization patterns support

zhongjuzhe committed:
add INT->FP conversion and FP->INT conversion auto-vectorization support

zhongjuzhe committed:
add auto-vectorization support: while_len. This is a special
  pattern for RVV loop handling

lhtin commented:
> See https://reviews.llvm.org/D120899 for right solution for this issue.

I am fixing in our inner GCC repo. I will update it when it is fixed.

zhongjuzhe committed:
add while_len support (including MD patterns)

zhongjuzhe committed:
add len_ load/store auto-vectorization support (including continues
  load store, gather/scatter, array load/store)(with mask or without
  mask)

zhongjuzhe committed:
add auto-vectorization vector permute support

zhongjuzhe committed:
add len_operation (add,sub,....) and len_shift auto-vectorization
  support

zhongjuzhe committed:
add len_conversion auto-vectorization support

zhongjuzhe committed:
add len comparision and len_vcond auto-vectorization support

zhongjuzhe committed:
add mixed operations

zhongjuzhe committed:
add vector-scalar mixed operation auto-vectorization support

zhongjuzhe committed:
add new target hook doc

zhongjuzhe committed:
add FMA missing information

zhongjuzhe committed:
add poly_uint_for_modes support.
  According to RVV spec, the maximum vector length is 65536 which
  will overflow in poly_int16, so we add poly_uint_for_modes support
  to support 65536 vector length

zhongjuzhe committed:
add COST model files (unfinished)

zhongjuzhe committed:
add rvv testcase including intrinsic+auto-vectorization

zhongjuzhe commented:
> See https://reviews.llvm.org/D120899 for right solution for this issue.

OK, we will fix it following this solution.

zhongjuzhe commented on gcc/config/riscv/riscv.md:
```diff
@@ -2051,7 +2051,59 @@
 {
-  if (riscv_expand_block_move (operands[0], operands[1], operands[2]))
+  if (TARGET_VECTOR && TARGET_RVV
+    && riscv_vector_expand_block_move (operands))
```
OK, I am agree with you. I am gonna move the `TARGET_VECTOR && TARGET_RVV` inside the function.
And also, I am so sorry that I missed the implementation. This patch is so huge that I always missed something when
I was splitting the codes.  
The implementation is here:  
https://github.com/riscv-collab/riscv-gcc/pull/334/commits/565fb69e77089a817f80e39af9403ff966002c41
Can you review this commit and give me more comments? I am gonna fix them including move `TARGET_VECTOR && TARGET_RVV` inside the implementation functions. Thank you~

zhongjuzhe replied:
I fix it here:https://github.com/riscv-collab/riscv-gcc/pull/334/commits/f43b675f0627890b0f004a77787d45799c67ac2b. Feel free to comment it.

zhongjuzhe committed:
fix bugs of tree-vect-stmts.cc

zhongjuzhe committed:
fix issues of riscv.h

zhongjuzhe commented on gcc/config/riscv/riscv.h:
```diff
@@ -287,11 +289,18 @@ ASM_MISA_SPEC
-	- FRAME_POINTER_REGNUM */
+	- FRAME_POINTER_REGNUM
+  - VL_REGNUM
+	- VTYPE_REGNUM
```
I have a fix here: https://github.com/riscv-collab/riscv-gcc/pull/334/commits/9c9dd3023c16e5a9dfb29490c5a9da3235997cc2. Feel free to comment it.

zhongjuzhe commented on gcc/config/riscv/riscv.h:
```diff
@@ -287,11 +289,18 @@ ASM_MISA_SPEC
+   - 32 vector registers:
+  - V0_REGS
+  - VNoV0_REGS
+  - V_REGS		*/
```
I fix it here: https://github.com/riscv-collab/riscv-gcc/pull/334/commits/9c9dd3023c16e5a9dfb29490c5a9da3235997cc2. Feel free to comment it.

zhongjuzhe commented on gcc/config/riscv/riscv.h:
```diff
@@ -376,6 +400,10 @@ ASM_MISA_SPEC
+#define RISCV_PROLOGUE_TEMP2_REGNUM (GP_TEMP_FIRST + 1)
+#define RISCV_PROLOGUE_TEMP2(MODE) gen_rtx_REG (MODE, RISCV_PROLOGUE_TEMP2_REGNUM)
+#define RISCV_PROLOGUE_TEMP3_REGNUM (GP_TEMP_FIRST + 1)
+#define RISCV_PROLOGUE_TEMP3(MODE) gen_rtx_REG (MODE, RISCV_PROLOGUE_TEMP3_REGNUM)
```
So far I didn't use `RISCV_PROLOGUE_TEMP3_REGNUM' so I remove it. The fix is here:https://github.com/riscv-collab/riscv-gcc/pull/334/commits/9c9dd3023c16e5a9dfb29490c5a9da3235997cc2. Feel free to comment it.

zhongjuzhe committed:
fix indent issues

zhongjuzhe commented on gcc/config/riscv/riscv.h:
```diff
@@ -450,6 +483,11 @@ enum reg_class
+  "VTYPE_REGS",								\
+  "V0_REGS", 				\
+  "VNoV0_REGS", 			\
+  "V_REGS", 							\
```
I fix indent issue here: https://github.com/riscv-collab/riscv-gcc/pull/334/commits/ddea6902b384bb8743ad576a3ee1d3d1d77fed4a. I just copied the white space from original riscv file. Is it OK for you?

zhongjuzhe replied:
Hi, I collected the format issues and indent issues. Then I fixed here: https://github.com/riscv-collab/riscv-gcc/commit/29e62a82471ceb3c915fc011c9f456e967035fe5. I have checked the riscv.h. The issue is fixed.
Feel free to comment.

zhongjuzhe commented on gcc/config/riscv/riscv.h:
```diff
@@ -466,13 +504,18 @@ enum reg_class
+  { 0x00000000, 0x00000000, 0x00000008, 0x00000000 },	/* VTYPE_REGS */	\
+  { 0x00000000, 0x00000000, 0x00000000, 0x00000001 },	/* V0_REGS */\
+  { 0x00000000, 0x00000000, 0x00000000, 0xfffffffe },	/* VNoV0_REGS */\
+  { 0x00000000, 0x00000000, 0x00000000, 0xffffffff },	/* V_REGS */\
```
I fix indent issue here: https://github.com/riscv-collab/riscv-gcc/pull/334/commits/ddea6902b384bb8743ad576a3ee1d3d1d77fed4a. I just copied the white space from original riscv file. Is it OK for you?

zhongjuzhe replied:
Hi, I collected the format issues and indent issues. Then I fixed here: https://github.com/riscv-collab/riscv-gcc/commit/29e62a82471ceb3c915fc011c9f456e967035fe5. I have checked the riscv.h. The issue is fixed.
Feel free to comment.

zhongjuzhe commented on gcc/config/riscv/riscv.h:
```diff
@@ -512,9 +555,17 @@ enum reg_class
+  104, 105, 106, 107, 108, 109, 110, 111,	\
+  112, 113, 114, 115, 116, 117, 118, 119, \
+  /* V0 ~ V7.  */					\
+  96,	97, 98, 99, 100, 101, 102, 103,	 \
```
I fix indent issue here: https://github.com/riscv-collab/riscv-gcc/pull/334/commits/ddea6902b384bb8743ad576a3ee1d3d1d77fed4a. I just copied the white space from original riscv file. Is it OK for you?

zhongjuzhe replied:
Hi, I collected the format issues and indent issues. Then I fixed here: https://github.com/riscv-collab/riscv-gcc/pull/334/commits/29e62a82471ceb3c915fc011c9f456e967035fe5. I have checked the `riscv.h`. The issue is fixed.
Feel free to comment.

zhongjuzhe committed:
remove useless comments and vsat vxrm

zhongjuzhe commented on gcc/config/riscv/riscv.h:
```diff
@@ -512,9 +555,17 @@ enum reg_class
   60, 61, 62, 63,							\
   /* Call-saved FPRs.  */						\
   40, 41, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59,			\
+  /* Call-clobbered vector registers, the order is consistent with Clang+LLVM Compiler */ \
```
I removed the comments. https://github.com/riscv-collab/riscv-gcc/pull/334/commits/1d4e28280190c4c076413881500b0177dfe98d4d. Feel free to comment it

zhongjuzhe commented on gcc/config/riscv/riscv.h:
```diff
@@ -780,7 +845,14 @@ typedef struct {
   "fa6", "fa7", "fs2", "fs3", "fs4", "fs5", "fs6", "fs7",	\
   "fs8", "fs9", "fs10","fs11","ft8", "ft9", "ft10","ft11",	\
-  "arg", "frame", }
+  "arg", "frame","vl","vtype","vxsat", "vxrm", "N/A", "N/A", \
```
I found so far I didn't use `vxsat` and `vxrm`. So I removed it: https://github.com/riscv-collab/riscv-gcc/pull/334/commits/1d4e28280190c4c076413881500b0177dfe98d4d. Feel free to comment it

zhongjuzhe committed:
fix cpymem strlen movstr cmpstr

zhongjuzhe commented on gcc/config/riscv/riscv.md:
```diff
@@ -2051,7 +2051,59 @@
+  ""
+{
+  if (TARGET_VECTOR && TARGET_RVV
+    && riscv_vector_expand_strlen (operands))
```
I fix it here:https://github.com/riscv-collab/riscv-gcc/pull/334/commits/f43b675f0627890b0f004a77787d45799c67ac2b. Feel free to comment it.

zhongjuzhe commented on gcc/config/riscv/riscv.md:
```diff
@@ -2051,7 +2051,59 @@
+  ""
+{
+  if (TARGET_VECTOR && TARGET_RVV
+    && riscv_vector_expand_strcpy (operands))
```
I fix it here:https://github.com/riscv-collab/riscv-gcc/pull/334/commits/f43b675f0627890b0f004a77787d45799c67ac2b. Feel free to comment it.

zhongjuzhe commented on gcc/config/riscv/riscv.md:
```diff
@@ -2051,7 +2051,59 @@
+  ""
+{
+  if (TARGET_VECTOR && TARGET_RVV
+    && riscv_vector_expand_strcmp (operands))
```
I fix it here:https://github.com/riscv-collab/riscv-gcc/pull/334/commits/f43b675f0627890b0f004a77787d45799c67ac2b. Feel free to comment it.

zhongjuzhe committed:
fix whitespace and code stype

zhongjuzhe commented on gcc/config/riscv/riscv-opts.h:
```diff
@@ -21,7 +21,8 @@ along with GCC; see the file COPYING3.  If not see
 
-enum riscv_abi_type {
+enum riscv_abi_type
+{
```
I fix the code style here: https://github.com/riscv-collab/riscv-gcc/pull/334/commits/15ba09286c0af8f6e23a2f9ece7a7a716002556b

zhongjuzhe commented on gcc/config/riscv/riscv.cc:
```diff
@@ -4569,7 +4569,7 @@ riscv_legitimize_move (machine_mode mode, rtx dest, rtx src)
       temp_reg = gen_reg_rtx (word_mode);
       zero_extend_p = (LOAD_EXTEND_OP (mode) == ZERO_EXTEND);
-      emit_insn (gen_extend_insn (temp_reg, src, word_mode, mode, 
+      emit_insn (gen_extend_insn (temp_reg, src, word_mode, mode,
```
It seems that the auto-trimming whitespace in my VScode create this issue. I fix it https://github.com/riscv-collab/riscv-gcc/pull/334/commits/15ba09286c0af8f6e23a2f9ece7a7a716002556b here.

zhongjuzhe committed:
fix whitespace

zhongjuzhe committed:
collect format and indent issues and fix them

zhongjuzhe commented on gcc/config/riscv/riscv.h:
```diff
@@ -430,6 +458,11 @@ enum reg_class
+  VTYPE_REGS,			/* vype register */
+  V0_REGS,		    /* v0.t registers */
+  VNoV0_REGS,		/* vector registers except v0.t */
+  V_REGS,			/* vector registers */
```
Hi, I collected the format issues and indent issues. Then I fixed here: https://github.com/riscv-collab/riscv-gcc/pull/334/commits/29e62a82471ceb3c915fc011c9f456e967035fe5. I have checked the `riscv.h`. The issue is fixed.
Feel free to comment.

zhongjuzhe committed:
remove redundant target hook definition

kito-cheng commented on gcc/config/riscv/vectorization.md:
```diff
@@ -5656,3 +5656,998 @@
+      (any_extend:<VW>
+        (match_operand:VWI 1 "register_operand"))
+      (any_extend:<VW>
+        (match_operand:VWI 2 "register_operand"))))]
```
You can just write all `match_operand` only for those pattern you will always expand, that should save some build time for gcc.
```
  [(match_operand:<VW> 0 "register_operand")
   (match_operand:VWI 1 "register_operand")
   (match_operand:VWI 2 "register_operand")]
```

zhongjuzhe replied:
> 

OK. I am gonna collect all cases to changed all of them as you suggested and make a single commit for you to review. Thanks~

zhongjuzhe replied:
> You can just write all `match_operand` only for those pattern you will always expand, that should save some build time for gcc.
> 
> ```
>   [(match_operand:<VW> 0 "register_operand")
>    (match_operand:VWI 1 "register_operand")
>    (match_operand:VWI 2 "register_operand")]
> ```

OK. I am gonna collect all cases to changed all of them as you suggested and make a single commit for you to review. Thanks~

zhongjuzhe replied:
> You can just write all `match_operand` only for those pattern you will always expand, that should save some build time for gcc.
> 
> ```
>   [(match_operand:<VW> 0 "register_operand")
>    (match_operand:VWI 1 "register_operand")
>    (match_operand:VWI 2 "register_operand")]
> ```

Hi, kito. I am trying to optimize the pattern as you suggested. But I have no idea how to do so.
For example:
(define_expand "vec_convert_v<su><optab>v<mode><vw>"
  [(set (match_operand:<VW> 0 "register_operand")
    (plus_minus_mult:<VW>
      (any_extend:<VW>
        (match_operand:VWI 1 "register_operand"))
      (any_extend:<VW>
        (match_operand:VWI 2 "register_operand"))))]

I used code iterator `plus_minus_mult` and `any_extend` to expand `<optab>` and `<su>. Do you want me remove the 
code iterators and split the pattern into several patterns?  Thank you~

kito-cheng replied:
For those kind of pattern I think it's fine to keep that :)

zhongjuzhe committed:
add no trapping floating-point comparision support

zhongjuzhe committed:
fix bugs

zhongjuzhe committed:
refine intrinsic definition macro

zhongjuzhe committed:
add nontuple segment load store support and testcase

zhongjuzhe committed:
merge and rename the testcase so that the test files are
  manageable and files are reduced.

zhongjuzhe committed:
add reduction auto-vectorization tests
  and fix bugs of reduction auto-vectorization

zhongjuzhe committed:
refine load store testcase

zhongjuzhe committed:
add widen reduction optimization and testcase

zhongjuzhe committed:
remove redundant cond_* vs patterns

kito-cheng commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+  }
+
+  unsigned
+  calcsew_lmul_ratio (unsigned int vsew_arg, unsigned int vlmul_arg) const
```
calc_sew_lmul_ratio

kito-cheng commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+  }
+
+  bool
+  getsew_lmul_ratio_only_p () const
```
get_sew_lmul_ratio_only_p

kito-cheng commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+  }
+
+  unsigned
+  calcsew_lmul_ratio () const
```
calc_sew_lmul_ratio

kito-cheng commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+	if (dump_file)
+	  fprintf (dump_file, "-------- Can't compare invalid VSETVLI Infos --------\n\n");
+	gcc_unreachable ();
+      }
```
Could you refine those checking code like:
```
    if (cond)
      {
	if (dump_file)
          fprintf (dump_file, DEBUG_MSG);
	gcc_unreachable ();
      }
```
into
```
    gcc_assert (cond && DEBUG_MSG);
```

kito-cheng commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+  }
+
+  void
+  operator= (const vinfo &other)
```
`vinfo &operator= (const vinfo &other)` and `return *this;` at end of function.

kito-cheng commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+};
+
+std::map<unsigned int, bb_vinfo> m_bb_vinfo_map;
+std::queue<basic_block> m_bb_queue;
```
`static` for `m_bb_vinfo_map` and `m_bb_queue`, and rename those two variable, `m_` means member data in general, but apparently that's global variable rather than member data.

kito-cheng commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+// Return a vinfo representing the changes made by this VSETVLI or
+// VSETIVLI instruction.
+static vinfo
+getinfoforvsetvli (rtx_insn *insn, vinfo curr_info)
```
get_info_for_vsetvli

kito-cheng commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+}
+
+static vinfo
+computeinfoforinstr (rtx_insn *insn, vinfo curr_info)
```
compute_info_for_instr

kito-cheng commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+}
+
+static bool
+canskipvsetvliforloadstore (rtx_insn *insn, vinfo &new_info, vinfo &curr_info)
```
can_skip_vsetvli_for_load_store_p

kito-cheng commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+}
+
+static bool
+needvsetvli (rtx_insn *insn, vinfo &new_info, vinfo &curr_info)
```
need_vsetvli

kito-cheng commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+}
+
+static bool
+needvsetvliphi (vinfo &new_info, rtx_insn *rtl)
```
need_vsetvli_phi

kito-cheng commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+        vse16_v_i16m1(c, vc, vl);
+
+    }
+
```
Could you remove redundant blank lines in example code?

kito-cheng commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+}
+
+static bool
+computevlvtypechanges (basic_block bb)
```
vl_vtype_changes_p

kito-cheng commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+}
+
+static void
+computeincomingvlvtype (const basic_block bb)
```
compute_incoming_vlvtype

kito-cheng commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+}
+
+static void
+insertvsetvli (rtx_insn *insn, const vinfo &curr_info, const vinfo &prev_info)
```
insert_vsetvli

kito-cheng commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+}
+
+static void
+emitvsetvlis (const basic_block bb)
```
emit_vsetvlis

zhongjuzhe replied:
> emit_vsetvlis

OK. Thanks for your suggestions. I am going to collect all your suggestions on revision and modify them into a single commit. So that I won't missed any information you gave.

zhongjuzhe replied:
> emit_vsetvlis

Hi, kito. I fixed insert vsetvlic PASS issue: https://github.com/riscv-collab/riscv-gcc/pull/334/commits/1bc0aa3c0319eb447468cdd074df80b4bc46267a. Thanks~

zhongjuzhe committed:
reconstitute the RVV functions framework: Move all the RVV releted
  functions from riscv.cc into riscv-vector.cc.

zhongjuzhe commented on gcc/config/riscv/riscv-modes.def:
```diff
@@ -20,3 +20,216 @@ along with GCC; see the file COPYING3.  If not see
+
+/* Half-precision floating point for __fp16.  */
+FLOAT_MODE (HF, 2, 0);
+ADJUST_FLOAT_FORMAT (HF, &ieee_half_format);
```
Yearh, you are right. I did missed something in some commits and make some commits mixed up. Sorry for that. Since the GCC was developed and it's painful to split them into several commits. Maybe in the next 1 ~ 2 weeks, after I finish the C++ overloaded functions. I am gonna split a clean GCC that only support RVV intrinsic (No any middle-end change) and test it fully. Then I make a new PR for it, at that time I will try my best to split them more reasonably for you to review (for example: zfh support, register spill.....etc). Before that, can you review the GCC codes base on this PR for the moment?  I am glad that you gave me suggestions.

zhongjuzhe commented on gcc/config/riscv/riscv.cc:
```diff
@@ -100,7 +100,7 @@ enum riscv_address_type {
 struct GTY(())  riscv_frame_info {
   /* The size of the frame in bytes.  */
-  HOST_WIDE_INT total_size;
+  poly_int64 total_size;
```
Agree. I will split it in the next PR.

zhongjuzhe commented on gcc/config/riscv/riscv.opt:
```diff
@@ -70,6 +70,67 @@ Enum(abi_type) String(lp64f) Value(ABI_LP64F)
+
+EnumValue
+Enum(riscv_vector_bits) String(65536) Value(RVV_65536)
+
```
Agree. I will split it in the next PR.

zhongjuzhe commented on gcc/config/riscv/riscv.cc:
```diff
@@ -1615,12 +1615,269 @@ riscv_legitimize_const_move (machine_mode mode, rtx dest, rtx src)
+          " option %<-march%>, or by using the %<target%>"
+          " attribute or pragma");
+  reported_p = true;
+}
```
Agree. I will split it in the next PR.

zhongjuzhe committed:
fix riscv-insert-vsetvli.cc issues

zhongjuzhe commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+#define VSETVLI_INSN_P(INSN) \
+  INSN && INSN_P(INSN) \
+  && recog_memoized (INSN) >= 0 \
+  && get_attr_type (INSN) == TYPE_VSETVL
```
> I would suggest you can use function for that instead of marco.

I fixed this issue: https://github.com/riscv-collab/riscv-gcc/pull/334/commits/1bc0aa3c0319eb447468cdd074df80b4bc46267a. Thanks~

zhongjuzhe commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+/* if the insn use vtype reg should always be the last
+   of a parallel. */
+#define USE_VTYPE_P(INSN) \
+  use_vtype_p (INSN)
```
I fixed this issue: https://github.com/riscv-collab/riscv-gcc/pull/334/commits/1bc0aa3c0319eb447468cdd074df80b4bc46267a. Thanks~

zhongjuzhe commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+/* if the insn use vl reg should always be the second
+   of a parallel. */
+#define USE_VL_P(INSN) \
+  use_vl_p (INSN)
```
I fixed this issue: https://github.com/riscv-collab/riscv-gcc/pull/334/commits/1bc0aa3c0319eb447468cdd074df80b4bc46267a. Thanks~

zhongjuzhe commented on gcc/config/riscv/riscv-insert-vsetvli.cc:
```diff
@@ -0,0 +1,1853 @@
+  }
+
+  void
+  setunknown ()
```
I fixed this issue: https://github.com/riscv-collab/riscv-gcc/pull/334/commits/1bc0aa3c0319eb447468cdd074df80b4bc46267a. Thanks~

zhongjuzhe committed:
rewrite RVV machine description files
  vector.md (base RVV)
  vector-auto.md (auto-vectorization)
  vector-opt.md (optimization)

zhongjuzhe committed:
rewrite the vcond auto-vectorization

zhongjuzhe committed:
remove redundant constraint

majin2020 commented on gcc/config/riscv/riscv.cc:
```diff
@@ -4593,18 +4613,45 @@ riscv_emit_stack_tie (void)
 }
 
+static void
+riscv_adjust_vector_frame (rtx target, poly_int64 offset)
```
Hi @zhongjuzhe  ,
  I have some questions about the frame_info and stack  of variable length type for your help. Could you give me some advice on this?
In the current gcc framework, it seems that 'poly_uint64(coeff0, coeff1)' is used to calculate 'cfun->machine->frame', but only one variable length type(vector) can be processed.
  If I want to add a new variable length type, how can I handle its frame_info and stack? In other words, how to handle the frame_info and stack of various variable length MACHINE_MODE.
  Thanks, looking forward to your comments.

zhongjuzhe replied:
I am not sure I am on the same page with you.
Are you asking how to handle the randomly various size of stack that includes different size of scalable vector.
If so, the answer is:
the  "poly_int64 offset" is the total size of the stack we need to assign.
For example, in "rvv-next" (not the same as upstream GCC that I support), 
"vint32m1_t" vector type has the size of [16,16], "vint32mf2_t" has vector type has the size of "[8,8].

If you want to assign a stack for for example: 2 "vint32m1_t" and 3 "vint32mf2_t" and additionl what  ever bytes (let's say 17 bytes). Here the total size of stack = 2 *[16,16] + 3 * [8,8] + 17 bytes. Then the poly_int64 offset = [73,56].  
riscv_adjust_vector_frame will handle this size of stack using csrr vlenb + scalar instructions to calculate the offset, and then assign this stack by sp - offset. Now riscv_adjust_vector_frame can handle any combination of poly_int64 value.

If you want to add a new variable length MACHINE_MODE that this rvv-next doesn't have, For example: VNx2TImode.
You just need to add it and use ADJUST_BYTES it correctly.  For example, suppose the VNx2TI occupy 2 RVV vector registers. You can just add this mode in riscv-modes.def:
VECTOR_MODE (VNx2TI)
ADJUST_BYTES (VNx2TI, poly_uint16 (16 * 2, 16 * 2) ([16,16]denotes a single vector register)

That's enough.Then no more work need to be done.

majin2020 replied:
Thank you for your quick and detailed reply, which has greatly benefited me.

  I understand the function riscv_adjust_vector_frame according to your reply, but there is still a question. Can you help me again?

  I want to add a new variable length "matrix" register, similar to the vector register, which's MACHINE_MODE  is VECTOR_MODE_P. Similar to "csrr vlenb", "csrr mlenb" is used. The mlenb represents the length of the matrix register.  Everything goes smoothly. But when matrix and vector are used together, it is difficult to distinguish when to use "csrr vlenb" and when to use "csrr mlenb" in  the function riscv_adjust_vector_frame, mainly because there is no relationship between mlenb and vlenb, and the poly_int64 offset seems unable to represent more than one variable length type. 

 I tried several times, but still couldn't find a good solution. Do I have a chance to get your comments again? Or if you have any doubts about my question, I can add them.

  Thanks.

zhongjuzhe replied:
OK. I understand you want to support your private "Matrix" extension in RVV GCC.

1. To design your own "Matrix" extension base on RVV. I think you should study and reference to"SME" (ARM Matrix extension). It's easily obtained in google search. Their ISA design is nice and clean. I would suggest you study it and redesign "Matrix" extension in RISC-V following "SME". But I could tell your basic ideas about "SME":
    -  SME is base on SVE.  Same, your can design RISC-V Matrix extension (I call it RVM may be). RVM base RVV.
    -  When SME is enabled, SVE is enabled too by default. Same to RVM and RVV.
    -  Matrix register size in SME is strictly related to vector length of SVE. The Matrix register size = row size * 
       column size. 
    -  The relationship between SME matrix register size and SVE vector register size:
        row size of "Matrix" register = column size of "Matrix" register = Machine vector length of SVE vector.
        So by copying the idea of SME, I suggest you define your Matrix register size = vlenb * vlenb in RVV.
   -   Base on the assumption I listed above, there is no need to introduce a new instruction "csrr mlenb". Because 
       "csrr vlenb" can also handle "Matrix" register variable length. After you studied ISA of SME, you will learn my
       idea. 

     So I highly recommand you redesign your Matrix extension following SME. 
     Let's say define matrix register size = vlenb * vlenb. Then add a instruction like matmul. 
     Define a matrix machine mode in riscv-modes.def: VECTOR_MODE (Matrix mode).
     However, the size of matrix is complicated.
    You should ADJUST_BYTES (256, 512, 256) to represent 2 runtime invariant.
     To handle this matrix register, you don't need distinguish them with RVV. They can be put together be handled.
      The current RVV GCC can not handle poly with 3 coeffs that for example assign a stack for  2 Matrix register + 3 
      vector register + 13 bytes. To handle that, the instructions sequence will be : 
      csrr a5, vlenb
      mult a6, a5, a5
      mult a5, a5, 3
      add  a5, a6, a5
      addi a5, a5, 13
      Then the value of a5 is the total stack size that you need to assign. Also to speed up this sequence, you could 
      introduce "csrr a5,mlenb" and revise poly handling in RVV GCC,  
     "csrr a5,mlenb" shoule be equal to "csrr a5,vlenb + mult a5,a5,a5".
     
2.  The size of matrix register can not reuse the poly framework of RVV. 
    Because the runtime invariant of RVV is 16 + 16x1, so in RVV GCC I define "#define NUM_POLY_INT_COEFFS 2".
    For a single vector register, size = (16,16). runtime invariant = 16 + 16*x1.
     
   However, for matrix, it's more complicated. Your should "#define NUM_POLY_INT_COEFFS 3".
   The poly value will become poly (coeffs0, coeffs1, coeffs2)
   The runtime invariant = (16+16*x1) * (16+16*x1) = 256 + 512*x1 + 256*x2^2 to represent a matrix register size.
    For a single RVV vector reigster, the runtime invariant = (16+16 * x1) * (16 + 16 * 0) = original poly value.
    To learn more about runtime invariant in GCC, I suggest you read this:
    https://gcc.gnu.org/onlinedocs/gccint/poly_005fint.html#poly_005fint
3. Base on the information above, you can know if we define both row and column are vector-length agnostic.
    You should refactor the poly handling in RVV, you can not reuse that. It may be a huge amount of work.
    To simplify this, may be you could define matrix reg row as vector-length agnostic and column as a compile-time
    constant, it will be easier and you could reuse the poly handling of RVV.

majin2020 replied:
Thank you for your suggestion. 

Since I want the matrix extension to be completely decoupled from the vector extension, I do not plan to follow the SME approach. 

I am using "#define NUM_POLY_INT_COEFFS 3" to achieve my goal. It is basically OK now.  As you said, I need to refactor the poly handling, but since poly may have taken into account the subsequent expansion when it was designed, the workload does not seem to be as large as expected.

Thanks again :)

ijinma commented:
Dear zhe @zhongjuzhe 

Thank you for your guidance. I have completed the "Matrix" extension in RVV GCC. Thank you again.

Recently, I have been very interested in the vector extension of RVV GCC and have studied your code seriously.

Although I have only joined RISCV for a short time,  I am very willing to contribute to the community. Is there any remaining work I can do about RVV GCC?

zhongjuzhe commented:
> 

Sure, we have a lot of stuff need to do. So far I am developing && maintaining RVV support in 2 different directories.
1). The RVV support in GCC upstream: https://github.com/gcc-mirror/gcc.
2). The RVV support in RISC-V foundation which you use: https://github.com/riscv-collab/riscv-gcc/tree/riscv-gcc-rvv-next.

The feature of RVV in upstream GCC is far behind rvv-next. 
I am working on cleaning up the codes belongs to rvv-next and pushing them GCC upstream.
So far about 10 patches are already merged into upstream GCC.

Hopefully, the upstream GCC will support RVV intrinsics at the end of this year and I am gonna support autovectorization in the
next year for upstream GCC.

1. For upstream RVV GCC support, I think you can help for review process:
https://gcc.gnu.org/pipermail/gcc-patches/2022-October/thread.html 

But don't push any codes related to RVV because you may interference my current work.

2. For rvv-next, we have a lot of stuff to do, you can develop it first and then make a PR:
    https://github.com/riscv-collab/riscv-gcc/pulls and call me to review and merge. Such codes will be eventually merged into
    upstream GCC.

3. TODO List of rvv-next:
    1. dynamic choosing best LMUL for auto-vectorization (so far compiler choose LMUL=1 to vectorize codes by default. But
        we should be able to choose LMUL = 2, 4, or 8 to vectorize if it is optimal).
    2. RTX cost model for poly handling (In riscv_const_insns function, see "case CONST_POLY_INT" case).
    3. RTX cost model for RVV instructions (In riscv_rtx_costs function).
    4. auto-vectoriztion cost model (TARGET_VECTORIZE_CREATE_COSTS and 
        TARGET_VECTORIZE_BUILTIN_VECTORIZATION_COST).
    5. More autovectorization testcases and support: For example, we don't support vectorization of floating point round:
        for (int i = 0; i < n; i++)
          a[i] = rond (b[i]);  RVV GCC can not vectorize such cases.

xl0714 commented:
Hi @zhongjuzhe .
When will the full vector intrinsic api be supported on  (https://github.com/gcc-mirror/gcc) ?
Can 90% be supported by the end of March? 
Look forward to your reply. 
Thank you.

zhongjuzhe commented:
Currently, upstream GCC support all RVV intrinsics except vleff and segment loads/stores.

I will support vleff today and it will be merged soon.

After vleff finished, we will no longer support any more features until GCC 13 release (End of April).
During this time from the week to GCC 13 release, we (Rivai && sifive) will keep testing && fix bugs && refine codes  until GCC 13 release.

So upstream GCC will support all RVV intrinsics except segment instructions at the end of March.

I am going to support segment loads/stores && autovectorization when GCC 14 is open (Hopefully in May).

Thanks.


***@***.***
 
From: xl0714
Date: 2023-03-06 09:04
To: riscv-collab/riscv-gcc
CC: Ju-Zhe Zhong; Mention
Subject: Re: [riscv-collab/riscv-gcc] PR for RVV support using splitted small chunks (PR #334)
Hi @zhongjuzhe .
When will the full vector intrinsic api be supported on (https://github.com/gcc-mirror/gcc) ?
Can 90% be supported by the end of March?
Look forward to your reply.
Thank you.
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***>

xl0714 commented:
Hi, @zhongjuzhe .
Can all vector intrinsic api be supported in GCC 13 release ？If not, how much work is left?
Look forward to your reply. 
Thank you.

zhongjuzhe commented:
> 

Only segment loads/stores are not supported in GCC 13 release since the api is still under discussions.
You can comment this PR if you are interested at it.
https://github.com/riscv-non-isa/rvv-intrinsic-doc/pull/198

xl0714 commented:
Thank you for your quick and detailed reply, which has greatly benefited me.
